### PR TITLE
SSTTP-1176 - Upgrade service to Play 2.5.12

### PR DESCRIPTION
**Pull request by cavster**

### CODE REVIEW

File: app/uk/gov/hmrc/timetopay/arrangement/config/JurisdictionCheckerConfig.scala [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.timetopay.arrangement.config
+
+import play.api.Configuration
+
+
+case class JurisdictionCheckerConfig(stotishPrefix:String , welsePrefix:String)
```
Comment (Smudgey) [inline app/uk/gov/hmrc/timetopay/arrangement/config/JurisdictionCheckerConfig.scala]: Typos - stotishPrefix, welsePrefix

File: app/uk/gov/hmrc/timetopay/arrangement/config/microserviceWiring.scala [hunk 2]
```diff
@@ -23,11 +23,11 @@ import uk.gov.hmrc.play.audit.http.config.LoadAuditingConfig
 import uk.gov.hmrc.play.audit.http.connector.AuditConnector
 import uk.gov.hmrc.play.auth.microservice.connectors.AuthConnector
 import uk.gov.hmrc.play.config.{AppName, RunMode, ServicesConfig}
-import uk.gov.hmrc.play.http.{HttpPost, HttpGet}
+import uk.gov.hmrc.play.http.{HttpGet, HttpPost}
 import uk.gov.hmrc.play.http.hooks.HttpHook
 import uk.gov.hmrc.play.http.ws._
 import uk.gov.hmrc.timetopay.arrangement._
-import uk.gov.hmrc.timetopay.arrangement.services._
+import uk.gov.hmrc.timetopay.arrangement.services.{DesTTPArrangementBuilder, _}
```
Comment (Smudgey) [inline app/uk/gov/hmrc/timetopay/arrangement/config/microserviceWiring.scala]: Optimize this import please

File: app/uk/gov/hmrc/timetopay/arrangement/config/microserviceWiring.scala [hunk 3]
```diff
@@ -68,10 +68,13 @@ trait ServiceRegistry extends ServicesConfig {
   lazy val arrangementDesApiConnector = DesArrangementApiService
 
   import play.api.Play.current
+  lazy val JurisdictionCheckerService = new JurisdictionChecker(JurisdictionCheckerConfig.create(configuration.getConfig("jurisdictionChecker")
+    .getOrElse(throw new RuntimeException("LetterAndControl configuration required"))))
+
   lazy val letterAndControlService = new LetterAndControlBuilder(LetterAndControlConfig.create(configuration.getConfig("letterAndControl")
-    .getOrElse(throw new RuntimeException("LetterAndControl configuration required")))) {}
+    .getOrElse(throw new RuntimeException("LetterAndControl configuration required"))),JurisdictionCheckerService) {}
```
Comment (Smudgey) [inline app/uk/gov/hmrc/timetopay/arrangement/config/microserviceWiring.scala]: Double check this as discussed

File: app/uk/gov/hmrc/timetopay/arrangement/services/DesTTPArrangementBuilder.scala [hunk 4]
```diff
@@ -23,7 +23,7 @@ import uk.gov.hmrc.timetopay.arrangement._
 import uk.gov.hmrc.timetopay.arrangement.services.JurisdictionType.{JurisdictionType, Scottish}
 
 
-class DesTTPArrangementBuilder {
+class DesTTPArrangementBuilder(jurisdictionChecker: JurisdictionChecker) {
```
Comment (Smudgey) [inline app/uk/gov/hmrc/timetopay/arrangement/services/DesTTPArrangementBuilder.scala]: Anyway to make this implicit / not have to pass this through everywhere?

File: project/MicroServiceBuild.scala [hunk 5]
```diff
@@ -11,32 +8,32 @@ object MicroServiceBuild extends Build with MicroService {
...
+  private val playConfigVersion = "4.1.0"
+  private val playJsonLoggerVersion = "3.1.0"
+  private val domainVersion = "4.1.0"
   private val hmrcTestVersion = "1.9.0"
   private val scalaTestVersion = "2.2.6"
   private val pegdownVersion = "1.6.0"
 
   private val playReactivemongoVersion = "4.8.0"
-
+  //todo hook em back in
```
Comment (Smudgey) [inline project/MicroServiceBuild.scala]: remove comment?

**Review by Smudgey — CHANGES_REQUESTED** [review]
Commit messages could also do with being updated / give more context for the changes

File: test/uk/gov/hmrc/timetopay/arrangement/services/DesArrangementServiceSpec.scala [hunk 6]
```diff
@@ -16,6 +16,8 @@
 
 package uk.gov.hmrc.timetopay.arrangement.services
 
+import javax.inject.Inject
```
Comment (Smudgey) [inline test/uk/gov/hmrc/timetopay/arrangement/services/DesArrangementServiceSpec.scala]: Can we remove this import?

File: test/uk/gov/hmrc/timetopay/arrangement/services/LetterAndControlBuilderSpec.scala [hunk 7]
```diff
@@ -16,19 +16,24 @@
 
 package uk.gov.hmrc.timetopay.arrangement.services
 
+import javax.inject.Inject
```
Comment (Smudgey) [inline test/uk/gov/hmrc/timetopay/arrangement/services/LetterAndControlBuilderSpec.scala]: Is this one still needed?

File: test/uk/gov/hmrc/timetopay/arrangement/services/TTPArrangementServiceSpec.scala [hunk 8]
```diff
@@ -16,6 +16,8 @@
 
 package uk.gov.hmrc.timetopay.arrangement.services
 
+import javax.inject.Inject
```
Comment (Smudgey) [inline test/uk/gov/hmrc/timetopay/arrangement/services/TTPArrangementServiceSpec.scala]: Same again?

**Review by Smudgey — COMMENTED** [review]
Last things I can see

File: test/uk/gov/hmrc/timetopay/arrangement/services/DesArrangementServiceSpec.scala (hunk 6 above)
Comment (cavster) [reply test/uk/gov/hmrc/timetopay/arrangement/services/DesArrangementServiceSpec.scala]: We need this one ;)

File: app/Module.scala [hunk 9]
```diff
@@ -0,0 +1,29 @@
+/*
```
Comment (Smudgey) [inline app/Module.scala]: This file could probably be deleted

File: app/uk/gov/hmrc/timetopay/arrangement/TTPArrangementController.scala [hunk 10]
```diff
@@ -16,17 +16,19 @@
...
+
 import play.api.Logger
 import play.api.libs.json.Json.toJson
-import play.api.mvc.{Result, Action, RequestHeader}
+import play.api.mvc.{Action, RequestHeader, Result}
 import uk.gov.hmrc.play.microservice.controller.BaseController
 import uk.gov.hmrc.timetopay.arrangement.services.{DesApiException, TTPArrangementService}
 
-import scala.concurrent.{Future, ExecutionContext}
+import scala.concurrent.{ExecutionContext, Future}
```
Comment (Smudgey) [inline app/uk/gov/hmrc/timetopay/arrangement/TTPArrangementController.scala]: It's not an issue, but wondering why these imports are just being moved around :P

File: app/uk/gov/hmrc/timetopay/arrangement/TTPArrangementRepository.scala [hunk 11]
```diff
@@ -16,18 +16,22 @@
 
 package uk.gov.hmrc.timetopay.arrangement
 
+import javax.inject.Inject
+
+import com.google.inject.{ImplementedBy, Singleton}
```
Comment (Smudgey) [inline app/uk/gov/hmrc/timetopay/arrangement/TTPArrangementRepository.scala]: The imports and spacing in this file can be cleaned up

File: app/uk/gov/hmrc/timetopay/arrangement/config/microserviceGlobal.scala [hunk 12]
```diff
@@ -34,22 +41,33 @@ object AuthParamsControllerConfiguration extends AuthParamsControllerConfig {
...
   override def controllerNeedsLogging(controllerName: String) = ControllerConfiguration.paramsForController(controllerName).needsLogging
 }
 
-object MicroserviceAuthFilter extends AuthorisationFilter {
+object MicroserviceAuthFilter extends AuthorisationFilter   with MicroserviceFilterSupport {
   override lazy val authParamsConfig = AuthParamsControllerConfiguration
   override lazy val authConnector = MicroserviceAuthConnector
   override def controllerNeedsAuth(controllerName: String): Boolean = ControllerConfiguration.paramsForController(controllerName).needsAuth
 }
+import uk.gov.hmrc.play.config.ServicesConfig
```
Comment (Smudgey) [inline app/uk/gov/hmrc/timetopay/arrangement/config/microserviceGlobal.scala]: Imports and spacings can be cleaned up here again. Should also avoid adding imports in half way down the file

File: app/uk/gov/hmrc/timetopay/arrangement/config/microserviceGlobal.scala [hunk 13]
```diff
@@ -34,22 +41,33 @@ object AuthParamsControllerConfiguration extends AuthParamsControllerConfig {
...
+class  GuiceModule extends AbstractModule with ServicesConfig {
```
Comment (Smudgey) [inline app/uk/gov/hmrc/timetopay/arrangement/config/microserviceGlobal.scala]: Double space after class

File: app/uk/gov/hmrc/timetopay/arrangement/config/microserviceWiring.scala [hunk 14]
```diff
@@ -15,22 +15,18 @@
  */
```
Comment (Smudgey) [inline app/uk/gov/hmrc/timetopay/arrangement/config/microserviceWiring.scala:15]: Imports + Spacings again

File: app/uk/gov/hmrc/timetopay/arrangement/config/microserviceWiring.scala [hunk 15]
```diff
@@ -43,54 +39,51 @@ object MicroserviceAuthConnector extends AuthConnector with ServicesConfig {
...
-object DesTTPArrangementBuilder extends DesTTPArrangementBuilder {}
-
-object RepositoryConfig  {
-  private implicit val connection = {
-    import play.api.Play.current
-    ReactiveMongoPlugin.mongoConnector.db
+class LetterAndControlAndJurisdictionCHecker extends ServicesConfig {
+  def createLetterAndControlConfig:LetterAndControlConfig = {
+    LetterAndControlConfig(
+      getConfString("letterAndControl.salutation", "sfsds"),
```
Comment (Smudgey) [inline app/uk/gov/hmrc/timetopay/arrangement/config/microserviceWiring.scala]: We need to think of more appropriate second messages, 'sfsds' accidentally got left in

File: app/uk/gov/hmrc/timetopay/arrangement/config/microserviceWiring.scala [hunk 16]
```diff
@@ -43,54 +39,51 @@ object MicroserviceAuthConnector extends AuthConnector with ServicesConfig {
...
+      getConfString("jurisdictionChecker.welsh.postcode.prefix", "sfsds")
+    ))
   }
-  lazy val TTPArrangementRepository = new TTPArrangementRepository
-
 }
 
-trait ServiceRegistry extends ServicesConfig {
+trait ServiceRegistry extends ServicesConfig with MongoDbConnection{
+  val test = new LetterAndControlAndJurisdictionCHecker()
```
Comment (Smudgey) [inline app/uk/gov/hmrc/timetopay/arrangement/config/microserviceWiring.scala]: Rename from test

File: app/uk/gov/hmrc/timetopay/arrangement/services/DesArrangementService.scala [hunk 17]
```diff
@@ -45,17 +45,16 @@ trait DesArrangementService {
...
     Logger.debug(s"Header carrier ${hc.headers}")
-
     http.POST[DesSubmissionRequest, HttpResponse](s"$desArrangementUrl/$serviceUrl", desSubmissionRequest)
       .map(_ => {
-        Logger.info(s"Submission successful for '${taxpayer.selfAssessment.utr}'")
         Right(SubmissionSuccess())
       }).recover {
-      case e: Throwable => onError(e)
+      case e: Throwable =>
+        println("We have an error " + e)
```
Comment (Smudgey) [inline app/uk/gov/hmrc/timetopay/arrangement/services/DesArrangementService.scala]: println accidentally left in

File: app/uk/gov/hmrc/timetopay/arrangement/services/DesArrangementService.scala [hunk 18]
```diff
@@ -45,17 +45,16 @@ trait DesArrangementService {
```
Comment (Smudgey) [inline app/uk/gov/hmrc/timetopay/arrangement/services/DesArrangementService.scala]: Imports + spacings

File: app/uk/gov/hmrc/timetopay/arrangement/config/microserviceWiring.scala [hunk 19]
```diff
@@ -43,54 +39,51 @@ object MicroserviceAuthConnector extends AuthConnector with ServicesConfig {
...
   override val http: HttpGet with HttpPost = WSHttp
 }
 
-object DesTTPArrangementBuilder extends DesTTPArrangementBuilder {}
-
-object RepositoryConfig  {
-  private implicit val connection = {
-    import play.api.Play.current
-    ReactiveMongoPlugin.mongoConnector.db
+class LetterAndControlAndJurisdictionCHecker extends ServicesConfig {
```
Comment (Smudgey) [inline app/uk/gov/hmrc/timetopay/arrangement/config/microserviceWiring.scala]: Checker has a capitalised H

File: app/uk/gov/hmrc/timetopay/arrangement/config/microserviceWiring.scala [hunk 20]
```diff
@@ -43,54 +39,51 @@ object MicroserviceAuthConnector extends AuthConnector with ServicesConfig {
...
+      getConfString("letterAndControl.office.officeName2", "sfsds"),
+      getConfString("letterAndControl.office.officePostCode", "sfsds"),
+      getConfString("letterAndControl.office.officePhone", "sfsds"),
+      getConfString("letterAndControl.office.officeFax", "sfsds"),
+      getConfString("letterAndControl.office.officeOpeningHours", "sfsds")
+    )
+  }
+  def createJurisdictionCheckerConfig:JurisdictionChecker = {
+    new JurisdictionChecker(JurisdictionCheckerConfig(
+      getConfString("jurisdictionChecker.scottish.postcode.prefix ", "sfsds"),
```
Comment (Smudgey) [inline app/uk/gov/hmrc/timetopay/arrangement/config/microserviceWiring.scala]: sfsds again

File: app/uk/gov/hmrc/timetopay/arrangement/services/DesTTPArrangementBuilder.scala [hunk 21]
```diff
@@ -17,13 +17,16 @@
...
+import javax.inject.Inject
 
 import play.api.Logger
 import uk.gov.hmrc.timetopay.arrangement._
+import uk.gov.hmrc.timetopay.arrangement.config.LetterAndControlAndJurisdictionCHecker
 import uk.gov.hmrc.timetopay.arrangement.services.JurisdictionType.{JurisdictionType, Scottish}
 
 
-class DesTTPArrangementBuilder {
+class DesTTPArrangementBuilder @Inject()(l:LetterAndControlAndJurisdictionCHecker){
```
Comment (Smudgey) [inline app/uk/gov/hmrc/timetopay/arrangement/services/DesTTPArrangementBuilder.scala]: Perhaps rename val from 'l'

File: app/uk/gov/hmrc/timetopay/arrangement/services/JurisdictionChecker.scala [hunk 22]
```diff
@@ -16,23 +16,21 @@
 
 package uk.gov.hmrc.timetopay.arrangement.services
```
Comment (Smudgey) [inline app/uk/gov/hmrc/timetopay/arrangement/services/JurisdictionChecker.scala:18]: Imports & spacing

File: app/uk/gov/hmrc/timetopay/arrangement/services/LetterAndControlBuilder.scala [hunk 23]
```diff
@@ -16,19 +16,20 @@
...
-import scala.concurrent.Future
-import scala.concurrent.ExecutionContext.Implicits.global
 import scala.util.Try
 
-class LetterAndControlBuilder(letterAndControlConfig: LetterAndControlConfig) {
-
+class LetterAndControlBuilder @Inject()(letterAndControlAndJurisdictionCHecker:LetterAndControlAndJurisdictionCHecker)   {
   type AddressResult = (Address, Option[LetterError])
-
+  println("bangsdfsdfdsfs")
```
Comment (Smudgey) [inline app/uk/gov/hmrc/timetopay/arrangement/services/LetterAndControlBuilder.scala]: remove println's

File: app/uk/gov/hmrc/timetopay/arrangement/services/LetterAndControlBuilder.scala [hunk 24]
```diff
@@ -16,19 +16,20 @@
 
 package uk.gov.hmrc.timetopay.arrangement.services
 
+import javax.inject.Inject
+
 import play.api.Logger
 import uk.gov.hmrc.timetopay.arrangement._
-import uk.gov.hmrc.timetopay.arrangement.config.LetterAndControlConfig
+import uk.gov.hmrc.timetopay.arrangement.config.{LetterAndControlConfig, LetterAndControlAndJurisdictionCHecker}
```
Comment (Smudgey) [inline app/uk/gov/hmrc/timetopay/arrangement/services/LetterAndControlBuilder.scala]: Optimise imports

File: app/uk/gov/hmrc/timetopay/arrangement/services/LetterAndControlBuilder.scala [hunk 25]
```diff
@@ -41,6 +42,7 @@ class LetterAndControlBuilder(letterAndControlConfig: LetterAndControlConfig) {
   }
 
   def create(ttpArrangement: TTPArrangement): LetterAndControl =  {
+    println("I was called")
```
Comment (Smudgey) [inline app/uk/gov/hmrc/timetopay/arrangement/services/LetterAndControlBuilder.scala]: Remove println

File: app/uk/gov/hmrc/timetopay/arrangement/services/LetterAndControlBuilder.scala [hunk 26]
```diff
@@ -16,19 +16,20 @@
...
 import scala.util.Try
 
-class LetterAndControlBuilder(letterAndControlConfig: LetterAndControlConfig) {
-
+class LetterAndControlBuilder @Inject()(letterAndControlAndJurisdictionCHecker:LetterAndControlAndJurisdictionCHecker)   {
   type AddressResult = (Address, Option[LetterError])
-
+  println("bangsdfsdfdsfs")
+  val LetterAndControlConfig = letterAndControlAndJurisdictionCHecker.createLetterAndControlConfig
+  println("bang")
```
Comment (Smudgey) [inline app/uk/gov/hmrc/timetopay/arrangement/services/LetterAndControlBuilder.scala]: Remove println

File: app/uk/gov/hmrc/timetopay/arrangement/services/TTPArrangementService.scala [hunk 27]
```diff
@@ -18,35 +18,37 @@ package uk.gov.hmrc.timetopay.arrangement.services
 
 import java.time.LocalDateTime
 import java.util.UUID
+import javax.inject.Inject
 
 import play.api.Logger
 import uk.gov.hmrc.play.http.HeaderCarrier
 import uk.gov.hmrc.timetopay.arrangement._
+import uk.gov.hmrc.timetopay.arrangement.config.{DesArrangementApiService, ServiceRegistry}
```
Comment (Smudgey) [inline app/uk/gov/hmrc/timetopay/arrangement/services/TTPArrangementService.scala]: Optimise imports

File: app/uk/gov/hmrc/timetopay/arrangement/services/TTPArrangementService.scala [hunk 28]
```diff
@@ -18,35 +18,37 @@ package uk.gov.hmrc.timetopay.arrangement.services
...
       s"and PP '${arrangement.paymentPlanReference}'")
 
-    val letterAndControl = letterAndControlService(arrangement)
-    val desTTPArrangement = desTTPArrangementService(arrangement)
+    val letterAndControl = letterAndControlBuilder.create(arrangement)
+    val desTTPArrangement = desTTPArrangementBuilder.create(arrangement)
 
     val request: DesSubmissionRequest = DesSubmissionRequest(desTTPArrangement, letterAndControl)
-
+    println("requestrequest " + request)
```
Comment (Smudgey) [inline app/uk/gov/hmrc/timetopay/arrangement/services/TTPArrangementService.scala]: Remove println

File: conf/application.conf [hunk 29]
```diff
@@ -158,7 +139,23 @@ microservice {
     }
 
     services {
-
+        letterAndControl {
```
Comment (Smudgey) [inline conf/application.conf:142]: These should ideally be moved out of the microservice wrapper, but if this is needed to get it to work for now, go for it

File: test/uk/gov/hmrc/timetopay/arrangement/TTPArrangementControllerSpec.scala [hunk 30]
```diff
@@ -16,23 +16,26 @@
 
 package uk.gov.hmrc.timetopay.arrangement
```
Comment (Smudgey) [inline test/uk/gov/hmrc/timetopay/arrangement/TTPArrangementControllerSpec.scala:18]: Optimise imports

File: test/uk/gov/hmrc/timetopay/arrangement/services/DesTTPArrangementBuilderSpec.scala [hunk 31]
```diff
@@ -18,18 +18,25 @@ package uk.gov.hmrc.timetopay.arrangement.services
 
 import java.time.LocalDate
 
+import org.scalamock.scalatest.MockFactory
 import org.scalatest.concurrent.ScalaFutures
-import uk.gov.hmrc.play.test.{UnitSpec, WithFakeApplication}
+import org.scalatest.mock.MockitoSugar
 import org.scalatest.prop.TableDrivenPropertyChecks._
+import org.scalatestplus.play.OneAppPerSuite
+import uk.gov.hmrc.play.test.{UnitSpec, WithFakeApplication}
```
Comment (Smudgey) [inline test/uk/gov/hmrc/timetopay/arrangement/services/DesTTPArrangementBuilderSpec.scala]: Optimise imports

File: test/uk/gov/hmrc/timetopay/arrangement/services/LetterAndControlBuilderSpec.scala [hunk 32]
```diff
@@ -16,19 +16,31 @@
...
 
-class LetterAndControlBuilderSpec extends UnitSpec with WithFakeApplication with ScalaFutures {
+class LetterAndControlBuilderSpec extends UnitSpec with MockFactory   with ScalaFutures  {
 
   val letterAndControlConfig = LetterAndControlConfig("Dear", "XXXX","XXXX","XXXX","XXXX","XXXX","XXXX", "XXXX","XXXX")
-
-  val letterAndControlService = new LetterAndControlBuilder(letterAndControlConfig)
+  val  jurisdictionConfig = JurisdictionCheckerConfig("^(AB|DD|DG|EH|FK|G|HS|IV|KA|KW|KY|ML|PA|PH|TD|ZE)[0-9].*",
+    "^(LL|SY|LD|HR|NP|CF|SA)[0-9].*")
+  println("sadasdsad")
```
Comment (Smudgey) [inline test/uk/gov/hmrc/timetopay/arrangement/services/LetterAndControlBuilderSpec.scala]: remove println's

File: test/uk/gov/hmrc/timetopay/arrangement/services/LetterAndControlBuilderSpec.scala [hunk 33]
```diff
@@ -16,19 +16,31 @@
...
+import uk.gov.hmrc.play.test.UnitSpec
 import uk.gov.hmrc.timetopay.arrangement._
-import uk.gov.hmrc.timetopay.arrangement.resources._
+import uk.gov.hmrc.timetopay.arrangement.config.{JurisdictionCheckerConfig, LetterAndControlConfig, LetterAndControlAndJurisdictionCHecker}
 import uk.gov.hmrc.timetopay.arrangement.resources.Taxpayers._
-import org.scalatest.prop.TableDrivenPropertyChecks._
+import uk.gov.hmrc.timetopay.arrangement.resources._
 
-class LetterAndControlBuilderSpec extends UnitSpec with WithFakeApplication with ScalaFutures {
+class LetterAndControlBuilderSpec extends UnitSpec with MockFactory   with ScalaFutures  {
```
Comment (Smudgey) [inline test/uk/gov/hmrc/timetopay/arrangement/services/LetterAndControlBuilderSpec.scala]: extra spaces after MockFactory

File: test/uk/gov/hmrc/timetopay/arrangement/services/TTPArrangementServiceSpec.scala [hunk 34]
```diff
@@ -16,119 +16,102 @@
```
Comment (Smudgey) [inline test/uk/gov/hmrc/timetopay/arrangement/services/TTPArrangementServiceSpec.scala:16]: imports

**Review by Smudgey — APPROVED** [review]

**Pull request merged**